src/coreclr/jit/emitxarch.cpp

```diff
@@ -596,26 +611,48 @@ bool emitter::AreUpper32BitsSignExtended(regNumber reg)
 
     instrDesc* id = emitLastIns;
 
-    if (id->idReg1() != reg)
-    {
-        return false;
-    }
+    bool result = false;
 
-    // movsx always sign extends to 8 bytes. W-bit is set.
-    if (id->idIns() == INS_movsx)
-    {
-        return true;
-    }
+    emitPeepholeIterateLastInstrs([&](instrDesc* id) {
+        if (emitIsInstrWritingToReg(id, reg))
+        {
+            switch (id->idIns())
+            {
+                // Conservative.
+                case INS_call:
+                    return PEEPHOLE_ABORT;
 
-    // movsxd is always an 8 byte operation. W-bit is set.
-    if (id->idIns() == INS_movsxd)
-    {
-        return true;
-    }
+                case INS_movsx:
+#ifdef TARGET_64BIT
+                case INS_movsxd:
+#endif // TARGET_64BIT
+                    if ((size == EA_1BYTE) || (size == EA_2BYTE))
+                    {
+                        result = (id->idOpSize() <= size);
+                    }
+#ifdef TARGET_64BIT
```

Comment:
On 32-bit will `EA_4BYTE` ever be passed? If not, we don't technically don't need the ifdef?

Reply:
`size` can be `EA_4BYTE` on 32-bit - ifdef'ing this only for 64-bit just does less work on 32-bit.